c/tskit/tables.c

```diff
@@ -6931,42 +6996,38 @@ tsk_table_collection_dump_indexes(tsk_table_collection_t *self, kastore_t *store
 }
 
 static int TSK_WARN_UNUSED
-tsk_table_collection_load_indexes(tsk_table_collection_t *self)
+tsk_table_collection_load_indexes(tsk_table_collection_t *self, kastore_t *store)
 {
     int ret = 0;
-    int has_insertion;
-    int has_removal;
-    const char *insertion_order = "indexes/edge_insertion_order";
-    const char *removal_order = "indexes/edge_removal_order";
+    tsk_id_t *edge_insertion_order = NULL;
+    tsk_id_t *edge_removal_order = NULL;
+    tsk_size_t num_rows;
+
     read_table_col_t read_cols[] = {
-        { insertion_order, (void **) &self->indexes.edge_insertion_order,
-            &self->edges.num_rows, 0, KAS_INT32, 0 },
-        { removal_order, (void **) &self->indexes.edge_removal_order,
-            &self->edges.num_rows, 0, KAS_INT32, 0 },
+        { "indexes/edge_insertion_order", (void **) &edge_insertion_order, &num_rows, 0,
+            KAS_INT32, TSK_COL_OPTIONAL },
+        { "indexes/edge_removal_order", (void **) &edge_removal_order, &num_rows, 0,
+            KAS_INT32, TSK_COL_OPTIONAL },
     };
 
-    has_insertion = kastore_containss(self->store, insertion_order);
-    if (has_insertion < 0) {
-        ret = has_insertion;
+    ret = read_table_cols(store, read_cols, sizeof(read_cols) / sizeof(*read_cols));
+    if (ret != 0) {
         goto out;
     }
-    has_removal = kastore_containss(self->store, removal_order);
-    if (has_removal < 0) {
-        ret = has_removal;
+    if ((edge_insertion_order == NULL) != (edge_removal_order == NULL)) {
+        ret = TSK_ERR_FILE_FORMAT;
         goto out;
     }
-
-    if (!has_insertion && !has_removal) {
-        /* If neither key is present the table is unindexed. */
-        self->indexes.malloced_locally = true;
-        self->indexes.edge_insertion_order = NULL;
-        self->indexes.edge_removal_order = NULL;
-    } else {
-        /* If one or the other key is present it's an error which will be
-         * handled by read_table_cols. */
-        self->indexes.malloced_locally = false;
-        ret = read_table_cols(
-            self->store, read_cols, sizeof(read_cols) / sizeof(*read_cols));
+    if (edge_insertion_order != NULL) {
+        if (num_rows != self->edges.num_rows) {
+            ret = TSK_ERR_FILE_FORMAT;
+            goto out;
+        }
+        ret = tsk_table_collection_set_index(
+            self, edge_insertion_order, edge_removal_order);
+        if (ret != 0) {
+            goto out;
+        }
```

Comment:
These lines have no effect. Is their aim to prevent an unchecked `ret` if code is added after at a later date? The pattern in the rest of the file is not to have these, but I can see an argument for them.

Reply:
I'm glad you brought this up! I was thinking about this, and I think we should probably have explicit ret != 0 checks after every call even if it's at the end of the function. There's two good reasons for this:

1. As you say, if code gets added later we won't accidentally skip the check (and it's really easy to miss this in review, as the diffs often won't show enough context to spot the problem)
2. Also it helps us with figuring out where test coverage is missing. If we're testing error conditions on every other call except the last one, we could miss testing some important error conditions.

So, even though it's obviously stupid, I think it's the right thing to do. It might result in a tiny bit more code (but maybe not: the compiler might spot that it's redundant), but it's not important. Since you're in favour, I'm going to loop back and make the change on all the code I've just updated now.